### PR TITLE
Actually forward gas on contract call

### LINKED ISSUE #184: Actually forward gas on call
Actually forward the amount of gas on contract calls

### PR DESCRIPTION
**Pull request by MujkicA**
Closes #184

This PR adds the field `gas_limit` to `CallParameters`. `gas_limit` is forwarded to the `CALL` instruction when generating the script for performing a contract call.
The difference between transaction gas limit and call gas limit is that the former applies to costs of the whole transaction while the latter specifies what portion of the transaction cost can be spent on the actual call.

Here is an example where we set a call gas limit of 1000
```rust

let call_params = CallParameters::new(None, None, Some(1000));

let result = contract_instance
        .initialize_counter(42)
        .call_params(CallParameters::new(call_params))
        .call()
        .await
        .expect_err("should error");

```

The new test `test_call_param_gas_errors` confirms that errors are reported correctly when setting transaction/call gas limits.

I've also amended the part on `CallParameters` in the docs. We should be able to just copy this in the appropriate section once the new docs structure is up.

### CODE REVIEW

**Review by Salka1988 — APPROVED** [review]

**Review by segfault-magnet — DISMISSED** [review]

**Review by vnepveu — CHANGES_REQUESTED** [review]
Some changes needed, thanks for fixing such a longstanding issue!

File: docs/src/getting-started/basics.md
Comment (adlerjohn) [reply docs/src/getting-started/basics.md:178]: Yes, changed in #384. All the more reason to include code instead of writing it inline!

File: packages/fuels-contract/src/script.rs
Comment (MujkicA) [reply packages/fuels-contract/src/script.rs:217]: The length of each parameter stored in data is in the docstring, I'll just write that we are getting an offset after those values

File: docs/src/getting-started/basics.md
Comment (MujkicA) [reply docs/src/getting-started/basics.md]: How can `CallResponse<D>` with a generic be added to the test function signature as output? I can't find any examples of this in code.

File: docs/src/getting-started/basics.md
Comment (MujkicA) [reply docs/src/getting-started/basics.md]: What would be the best way to include this since its outside of our examples?

Comment (vnepveu) [reply docs/src/getting-started/basics.md]: you can just add the `ANCHOR: xxx` tag as you would for code exemple then include the file. Just rebase/merge master, I included this already in my work on  docs migration.

File: docs/src/getting-started/basics.md
Comment (vnepveu) [reply docs/src/getting-started/basics.md]: Don't return this as an output, basically use `Result<(), Error>` as function return, and end your test function with `Ok(())`

File: docs/src/getting-started/basics.md
Comment (adlerjohn) [inline docs/src/getting-started/basics.md:163]:
```suggestion
The `gas_limit` parameter differs from the one in `TxParameters` in that it defines the limit for the actual contract call as opposed to the gas limit for the whole transaction. This means that it is constrained by the transaction limit. If it is set to an amount grater than the available gas, all available gas will be forwarded.
```

File: docs/src/getting-started/basics.md
Comment (adlerjohn) [inline docs/src/getting-started/basics.md:166]:
```suggestion
Here we use both the transaction and call `gas_limit` to specify that the contract call transaction may consume up to 1000 gas while the actual call may only consume up to 200 gas:
```

File: packages/fuels-abigen-macro/tests/test_projects/msg_amount/Cargo.toml
Comment (adlerjohn) [inline packages/fuels-abigen-macro/tests/test_projects/msg_amount/Cargo.toml:4]:
```suggestion
authors = ["Fuel Labs <contact@fuel.sh>"]
```

File: packages/fuels-abigen-macro/tests/test_projects/msg_amount/Forc.toml
Comment (adlerjohn) [inline packages/fuels-abigen-macro/tests/test_projects/msg_amount/Forc.toml:2]:
```suggestion
authors = ["Fuel Labs <contact@fuel.sh>"]
```

File: packages/fuels-contract/src/script.rs
Comment (adlerjohn) [inline packages/fuels-contract/src/script.rs:241]: ?
```suggestion
    /// 0x11 Gas limit
```

**Review by adlerjohn — COMMENTED** [review]
Avoid manual line breaks for markdown text.

**Review by vnepveu — APPROVED** [review]
lgtm

**Review by adlerjohn — COMMENTED** [review]
Looks good to me, will defer to @vnepveu or @digorithm for final review. Good work!

File: docs/src/getting-started/basics.md
Comment (digorithm) [inline docs/src/getting-started/basics.md:161]:
```suggestion
The `gas_limit` parameter differs from the one in `TxParameters` in that it defines the limit for the actual contract call as opposed to the gas limit for the whole transaction. This means that it is constrained by the transaction limit. If it is set to an amount greater than the available gas, all available gas will be forwarded.
```

**Review by digorithm — CHANGES_REQUESTED** [review]
LGTM, great stuff! Hopefully the tx's `gas_limit` and call's `gas_limit` won't be confusing to the users 🤔. 

Found just a small typo; feel free to merge after!

**Comment by adlerjohn** [comment]
> Hopefully the tx's gas_limit and call's gas_limit won't be confusing to the users 🤔.

I went back and forth on this and decided against commenting it, but then @digorithm brought it up so here goes: can @MujkicA can you file an issue to follow-up on this PR to change the call gas limit to something like "gas forwarded"? Semantically, a transaction has a limit to how much gas it can use, but isn't _forwarded_ any gas. When contract calls are made, gas is actually sent (forwarded) with the call.

**Comment by digorithm** [comment]
> > Hopefully the tx's gas_limit and call's gas_limit won't be confusing to the users 🤔.
> 
> I went back and forth on this and decided against commenting it, but then @digorithm brought it up so here goes: can @MujkicA can you file an issue to follow-up on this PR to change the call gas limit to something like "gas forwarded"? Semantically, a transaction has a limit to how much gas it can use, but isn't _forwarded_ any gas. When contract calls are made, gas is actually sent (forwarded) with the call.

Haha sorry for the troubles @MujkicA; after reading the docs I felt a bit confused about the terminology, but after re-reading and reading some code it made sense, but there was some friction. So I'm guessing that could happen with the users as well. 

I like John's suggestion here, but as he said, can leave it for a follow-up PR.

**Comment by MujkicA** [comment]
No problem at all. I'd rather change it in this PR if that is ok.

**Review by digorithm — APPROVED** [review]
Yup, that change in the naming made things much clearer. LGTM!

**Comment by digorithm** [comment]
Damn, this `enabled auto-merge (squash)` thingy is scary; for a second I thought I pressed the merge button 😂.

**Pull request merged**
